src/core/ComboBox/ComboBox.tsx

```diff
@@ -380,84 +459,365 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
             if (isOpen) {
-              setIsOpen(false);
+              dispatch(['select']);
             } else {
-              inputRef.current?.focus();
+              dispatch(['open']);
+            }
+            break;
+          }
+          case 'Escape': {
+            event.preventDefault();
+            dispatch(['close']);
+            break;
+          }
+        }
+      },
+      [dispatch, isOpen, onKeyDownProp, optionsRef],
+    );
+
+    const handleFocus = React.useCallback(
+      (event: React.FocusEvent<HTMLInputElement>) => {
+        dispatch(['open']);
+        onFocusProp?.(event);
+      },
+      [dispatch, onFocusProp],
+    );
+
+    return (
+      <Input
+        ref={refs}
+        onKeyDown={handleKeyDown}
+        onFocus={handleFocus}
+        aria-activedescendant={
+          isOpen && focusedIndex != undefined && focusedIndex > -1
+            ? getIdFromIndex(focusedIndex)
+            : undefined
+        }
+        role='combobox'
+        aria-controls={isOpen ? `${id}-list` : undefined}
+        aria-autocomplete='list'
+        spellCheck={false}
+        autoCapitalize='none'
+        autoCorrect='off'
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxInput.displayName = 'ComboBoxInput';
+
+const ComboBoxPopover = React.forwardRef(
+  (
+    props: PopoverProps & { children: JSX.Element },
+    forwardedRef: React.RefObject<Element>,
+  ) => {
+    const { children, ...rest } = props;
+    const { isOpen } = useSafeContext(ComboBoxStateContext);
+    const dispatch = useSafeContext(ComboBoxActionContext);
+    const { inputRef, toggleButtonRef } = useSafeContext(ComboBoxRefsContext);
+
+    // sync internal isOpen state with user's visible prop
+    React.useEffect(() => {
+      dispatch([props.visible ? 'open' : 'close']);
+    }, [dispatch, props.visible]);
+
+    return (
+      <Popover
+        placement='bottom-start'
+        visible={isOpen}
+        onClickOutside={React.useCallback(
+          (_, { target }) => {
+            if (!toggleButtonRef.current?.contains(target as Element)) {
+              dispatch(['close']);
             }
-          }}
+          },
+          [dispatch, toggleButtonRef],
+        )}
+        animation='shift-away'
+        duration={200}
+        reference={inputRef}
+        ref={forwardedRef}
+        content={children}
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxPopover.displayName = 'ComboBoxPopover';
+
+const ComboBoxMenu = React.forwardRef(
+  (
+    props: Omit<MenuProps, 'onClick'> & React.ComponentPropsWithoutRef<'ul'>,
+    forwardedRef: React.RefObject<HTMLUListElement>,
+  ) => {
+    const { className, style, ...rest } = props;
+    const { minWidth, id } = useSafeContext(ComboBoxStateContext);
+    const { menuRef, optionsRef } = useSafeContext(ComboBoxRefsContext);
+
+    React.useEffect(() => {
+      if (menuRef.current != null) {
+        const items = menuRef.current.querySelectorAll('[data-iui-index]');
```

Comment:
What won't work in custom rendered items that are missing this attribute?

Reply:
keyboard navigation

Reply:
and how would i get original index from children if this attribute is not present?